controllers/s3/s3bucket.go

```diff
@@ -0,0 +1,265 @@
+	if err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, "cannot create instance")
+	}
+
+	bucket.Status.AtProvider.ObjectUserID = objectUser.ID
+	bucket.Status.AtProvider.Status = statusOnline
+
+	accessKey, secretKey, err := s3.GetKeys(objectUser)
+	if err != nil {
+		return resource.ExternalCreation{}, err
+	}
+	cn := resource.ConnectionDetails{
+		runtimev1alpha1.ResourceCredentialsSecretUserKey:     []byte(accessKey),
+		runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(secretKey),
+		runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(cloudscale.S3Endpoint),
+	}
+
+	return resource.ExternalCreation{ConnectionDetails: cn}, nil
+}
+
+// Update the existing external resource to match the specifications of our
+// managed resource. resource.ManagedReconciler only calls Update if Observe
+// reported that the external resource was not up to date.
+func (e *external) Update(ctx context.Context, mg resource.Managed) (resource.ExternalUpdate, error) {
+	bucket, ok := mg.(*storagev1alpha1.S3Bucket)
+	if !ok {
+		return resource.ExternalUpdate{}, errors.New(errNotInstance)
+	}
+	log.Info("Update", "bucket", bucket.Name)
+	objectUser, err := e.s3Client.CreateOrUpdateBucket(ctx, bucket.Status.AtProvider.ObjectUserID, getBucketName(bucket), bucket.Spec.ForProvider.CannedACL, bucket.Spec.ForProvider.Tags)
+	bucket.Status.AtProvider.ObjectUserID = objectUser.ID
+	return resource.ExternalUpdate{}, errors.Wrap(err, "cannot update instance")
+}
+
+// Delete the external resource. resource.ManagedReconciler only calls Delete
+// when a managed resource with the 'Delete' reclaim policy has been deleted.
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	bucket, ok := mg.(*storagev1alpha1.S3Bucket)
+	if !ok {
+		return errors.New(errNotInstance)
+	}
+	log.Info("Delete", "bucket", bucket.Name)
+	// Indicate that we're about to delete the instance.
+	bucket.Status.AtProvider.Status = statusDeleting
+
+	// Delete the instance.
+	err := e.s3Client.DeleteBucket(ctx, bucket.Status.AtProvider.ObjectUserID, getBucketName(bucket))
+	if err != nil && !s3.IsErrorNotFound(err) {
+		return errors.Wrap(err, "cannot delete instance")
+	}
+	return nil
+}
+
+// Generate bucket name based on the NameFormat spec value,
+// If name format is not provided, bucket name defaults to UID
+// If name format provided with '%s' value, bucket name will result in formatted string + UID,
+//   NOTE: only single %s substitution is supported
+// If name format does not contain '%s' substitution, i.e. a constant string, the
+// constant string value is returned back
+//
+// Examples:
+//   For all examples assume "UID" = "test-uid"
+//   1. NameFormat = "", BucketName = "test-uid"
+//   2. NameFormat = "%s", BucketName = "test-uid"
+//   3. NameFormat = "foo", BucketName = "foo"
+//   4. NameFormat = "foo-%s", BucketName = "foo-test-uid"
+//   5. NameFormat = "foo-%s-bar-%s", BucketName = "foo-test-uid-bar-%!s(MISSING)"
```

Comment:
I was hoping to deprecate this `NameFormat` pattern in favor of the new `ExternalName` annotation, which I see you're also using. `NameFormat` is an older pattern we used to generate random names for dynamically provisioned managed resources before `ExternalName` gave more control over naming.

One feature of the `ExternalName` pattern (that is under documented) is that we copy the external name annotation from the resource claim to any dynamically provisioned managed resource (you can see this in [`resource.ObjectMetaConfigurator`](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#ObjectMetaConfigurator). 

This means that presuming you use `meta.GetExternalName` to name your bucket:

* Managed resources will be named `claimnamespace-claimname-randomstring` by default, and thus your external resource (i.e. actual bucket) would be named `claimnamespace-claimname-randomstring`.
* If you annotate your claim with `crossplane.io/external-name: my-cool-bucket` your managed resource will still be named `claimnamespace-claimname-randomstring`, but it will be annotated as `crossplane.io/external-name: my-cool-bucket` and thus your external resource would be named `my-cool-bucket`.

Would this work for your controller, or do you feel there's value in falling back to the `NameFormat` name?